src/librustc_jobserver/lib.rs

```diff
@@ -0,0 +1,292 @@
+use std::sync::Arc;
+use std::sync::Condvar;
+use std::sync::Mutex;
+
+lazy_static! {
+    // We can only call `from_env` once per process
+
+    // Note that this is unsafe because it may misinterpret file descriptors
+    // on Unix as jobserver file descriptors. We hopefully execute this near
+    // the beginning of the process though to ensure we don't get false
+    // positives, or in other words we try to execute this before we open
+    // any file descriptors ourselves.
+    //
+    // Pick a "reasonable maximum" if we don't otherwise have
+    // a jobserver in our environment, capping out at 32 so we
+    // don't take everything down by hogging the process run queue.
+    // The fixed number is used to have deterministic compilation
+    // across machines.
+    //
+    // Also note that we stick this in a global because there could be
+    // multiple rustc instances in this process, and the jobserver is
+    // per-process.
+    static ref GLOBAL_CLIENT: Client = unsafe {
+        Client::from_env().unwrap_or_else(|| {
+            log::trace!("initializing fresh jobserver (not from env)");
+            let client = Client::new(32).expect("failed to create jobserver");
+            // Acquire a token for the main thread which we can release later
+            client.acquire_raw().ok();
+            client
+        })
+    };
+
+    static ref HELPER: Helper = Helper::new();
+}
+
+const EMPTY: usize = 0;
+const CARGO_REQUESTED: usize = 1;
+const MAKE_REQUESTED: usize = 2;
+static TOKEN_REQUESTS: AtomicUsize = AtomicUsize::new(0);
+
+fn should_notify() -> bool {
+    let value = TOKEN_REQUESTS.load(Ordering::SeqCst);
+    assert!(value != EMPTY, "jobserver must be initialized");
+    value == CARGO_REQUESTED
+}
+
+/// This will only adjust the global value to the new value of token_requests
+/// the first time it's called, which means that you want to make sure that the
+/// first call you make has the right value for `token_requests`. We try to help
+/// out a bit by making sure that this is called before any interaction with the
+/// jobserver (which usually happens almost immediately as soon as rustc does
+/// anything due to spawning up the Rayon threadpool).
+///
+/// Unfortunately the jobserver is inherently a global resource (we can't
+/// have more than one) so the token requesting strategy must likewise be global.
+///
+/// Usually this doesn't matter too much, as you're not wanting to set the token
+/// requests unless you're in the one-rustc-per-process model, and we help out
+/// here a bit by not resetting it once it's set (i.e., only the first init will
+/// change the value).
+pub fn initialize(token_requests: bool) {
+    lazy_static::initialize(&GLOBAL_CLIENT);
+    lazy_static::initialize(&HELPER);
+    let previous = TOKEN_REQUESTS.compare_and_swap(
+        EMPTY,
+        if token_requests { CARGO_REQUESTED } else { MAKE_REQUESTED },
+        Ordering::SeqCst,
+    );
+    if previous == EMPTY {
```

Comment:
This is a little worrisome to me, could we panic on double initialization instead of ignoring it? It seems like we should be calling this at most once throughout the lifetime of rustc itself.

Also if possible it'd be even better to wrap all this up in some sort of structure that we shove into a `Session`, or maybe a structure put into a thread local. There's a lot of global-ness about this module which would be great if we could encapsulate better.

Reply:
If you mean the specific `TOKEN_REQUESTS` global, we can't really do that. This is explicitly something that must be process-global, because the jobserver is a global resource. I guess it is *possible* to try to do both -- e.g., we thread the jobserver file descriptors into a rustc `-Cjobserver=<mode=legacy|new>,fd,fd` or something which then makes it a Session-local resource.

Reply:
Hm I don't really understand, `Session` is already "practically global" but represents a lot of "proof of one-time initialization". I'm wondering here if we can do the same for jobserver tokens, since we really shouldn't be initializing this more than once. If this is initialized twice I think we should at least panic rather than paper over it as it currently does.

Reply:
We currently intend that Session (and resources associated with it) can coexist, i.e., you can have two Sessions. But we cannot have two jobservers.

(I think I did update it to panic, though, I agree with that).